chapters/annotations.tex

```diff
@@ -383,20 +383,25 @@ \section{Annotations for Simulations}\label{annotations-for-simulations}
 \end{center}
 
 \begin{annotationdefinition}[experiment]
-% henrikt-ma 2021-04: Seems strange to allow 'experiment' completely without list of options -- what would it mean?
 \begin{synopsis}[grammar]\begin{lstlisting}
-"experiment"
-   [ "(" [ experimentOption { "," experimentOption } ] ")" ]
-
-experimentOption:
-   "StartTime" "=" [ "+" | "-" ] UNSIGNED-NUMBER
-   | "StopTime" "=" [ "+" | "-" ] UNSIGNED-NUMBER
-   | "Interval" "=" UNSIGNED-NUMBER
-   | "Tolerance" "=" UNSIGNED-NUMBER
+annotation "experiment" "(" experimentModifiers ")"
 \end{lstlisting}\end{synopsis}
 \begin{semantics}
-The \lstinline{experiment} annotation defines the default start time (\lstinline!StartTime!) in {[}s{]}, the default stop time (\lstinline!StopTime!) in {[}s{]}, the suitable time resolution for the result grid (\lstinline!Interval!) in {[}s{]}, and the default relative integration tolerance (\lstinline!Tolerance!) for simulation experiments to be carried out with the model or block at hand.
+The modifiers are for the following record.
+\begin{lstlisting}[language=modelica]
+record experiment
+  Real StartTime(unit = "s") = 0;
+  Real StopTime(unit = "s");
+  Real Interval(unit = "s");
+  Real Tolerance;
+end experiment;
+\end{lstlisting}
+The \lstinline!experiment! annotation defines the start time (\lstinline!StartTime!) in {[}s{]}, the stop time (\lstinline!StopTime!) in {[}s{]}, the suitable time resolution for the result grid (\lstinline!Interval!) in {[}s{]}, and the relative integration tolerance (\lstinline!Tolerance!) for simulation experiments to be carried out with the model or block at hand.
+When \lstinline!Interval! or \lstinline!Tolerance! is not provided, the tool is responsible for applying appropriate defaults.
 If \lstinline!StartTime! is not specified it is assumed to be \lstinline!0.0!.
+
+If \lstinline!StopTime! is set the model is required to be a simulation model.
```

Comment:
We need to consider the options here, probably bringing them up for an informed discussion at a phone meeting.

I would like this to be a valid non-simulation model:
```
partial model MicroSecondExampleBase "Base class for examples on the micro second time scale"
  annotation(StopTime = 10e-6);
end MicroSecondExampleBase;
```

Some candidate designs to be included in a poll:
- `StopTIme` implies _simulation model_ (can lead to the kind of error shown in `MicroSecondExampleBase`)
- _Simulation model_ implies `StopTime` (making `StopTime` part of the _simulation time_ definition definition)

Reply:
As indicated that it is a different issue and shouldn't block this one.

The problem, as previously, indicated is that realistically you often have cases like this (but with more components) - in particular I see this used more often than the MicroSecondExampleBase:
```
model TestBed
   replaceable Modelica.Blocks.Interfaces.SignalSource source;
   Plant plant;
equation
   connect(source.y, plant.u);
   annotation(experiment(StopTime=1e2));
end TestBed;
```
Clearly:
* StopTime is relevant for some such cases - but not all - since the StopTime might vary a bit depending on the source.
* The model isn't a simulation model itself, so the StopTime cannot be applied here.
* Note that for some templates it doesn't make sense to include the 
* We thus need to both have a way of marking actual simulation models in libraries, and way to indicate simulation settings that can be inherited.

Reply:
> As indicated that it is a different issue and shouldn't block this one.

I find it difficult to see how we can wrap up this PR without agreeing on some relation between presence of `StopTime` and _simulation model_.  Otherwise, we need to find a way to incorporate some of the changes without getting into the details of this (that is, saying things like "… is required to be a simulation model"). However, the question about `StopTime` might very well be the most important aspect of this PR, so I think it's worthwhile trying to find a solution.

I agree that we don't want the `TestBed` above to be a problem.  It clearly cannot be a simulation model, so we would be shooting ourselves in the foot if we say that `StopTime` implies being a simulation model.  The solution seems easy though, and comes down to relying on a working definition of _simulation model_.

We already have this:
> Such a class is allowed to be incomplete, and cannot be instantiated in a simulation model; useful, e.g., as a base class.

Our definition of simulation model could be made consistent with this rule, so that it is fulfilled by definition.  We could define _simulation model_ as a class meeting all of these requirements:
- model or block
- not partial
- no components of partial type
- has `StopTime`

It would also make sense to narrow down where the `experiment` annotation is defined:
> A model or block may have an \lstinline!experiment! annotation, defining the start time…

According to such a definition neither `TestBed` nor `MicroSecondExampleBase` would be an error, and we have the nice property that every simulation model has a `StopTime`.

By the way, this looks incomplete:
> Note that for some templates it doesn't make sense to include the

Reply:
> Our definition of simulation model could be made consistent with this rule, so that it is fulfilled by definition. We could define _simulation model_ as a class meeting all of these requirements:
> 
>  * model or block
>  * not partial
>  * no components of partial type
>  * has `StopTime`

In OpenModelica we use the `experiment` annotations to automatically test libraries, so for us it's very much preferable to be able to determine whether a model can be simulated or not "at a glance". Defining a simulation model as a model that among other things has no components of partial type would make that a lot more complicated since that would involve having to look up each components type.

So if we want non-simulation models to be allowed to have `experiment` annotations I think it would be better to have that as part of the annotation, like `experiment(Template = true)` or something. Or possibly by allowing partial models to have `experiment` annnotations, but I'm not sure if that covers all use cases.

Reply:
To me it also a matter of intent. If a model has components of a partial type, was it because someone forgot to redeclare it or is it because it was intended as a template. 
A human can see that if it is in "Examples"-package it is likely a missed redeclare, but in "Templates"-package it is likely intended as a template.

Reply:
> Defining a simulation model as a model that among other things has no components of partial type would make that a lot more complicated since that would involve having to look up each components type.

This is a good point, exactly the kind of thing we need to consider when working out a good definition of _simulation model_.  I agree that it would be great if _simulation model_ could be determined on a glance, but I guess it would be a routine operation for a Modelica tool to verify the condition.

I hope that we can do this without introducing a new annotation like `Template = true`, but it remains to figure out.  It would have been great if we could just have relied on the class itself being declared `partial` instead of adding `Template = true`, but I suppose one doesn't want to do it in a case like this as it unnecessarily restricts the ways in which the class can be instantiated in simulation models…

Reply:
We already have a good definition of simulation model:
A class (of specialized class model, class or block) intended to be simulated on its own is called a
simulation model.
That imposes various semantic restrictions and they are listed.

Reply:
The problem with that definition is that _intention_ is something subjective in the eyes of the modeler, and not something suitable for formal reasoning and implementation in tools.

Even though class validation is not something the specification deals with directly, I expect to find this kind of functionality in all Modelica tools: You simply as the tool if a given class fulfills requirements that can be checked on the class itself (without it being instantiated somewhere inside a simulation model).  When it comes to models and blocks, it would be great to have a definition of _simulation model_ that would unambiguously allow the tool to determine whether to check the full set of requirements applicable to a simulation model, or to just check the smaller set of requirements applicable to a non-simulation model.

To be of practical use, the definition needs to be simpler (easier to comprehend, quicker to check, etc) than the implicit definition that a model (block) is a simulation model iff and only if it fulfills everything needed to be simulated.

Basically I would expect that a tool only allows simulation models to be simulated.  However, for quick and dirty work, I would be OK if the tool allows me to request a model without `StopTime` to be simulated, but if I do so I should get informed that `StopTime` is required and be presented with some method of providing a value on the fly.  When requesting validation of the same model, I'd like to be informed that it is not a simulation model, along with any other information coming out of the non-simulation model validation.

A useful definition should also come with some way of turning a simulation model into a non-simulation model with minimal other side effects, so that a modeler that doesn't want the model to be validated against the full set of requirements can simply opt out of being possible to simulate.  For example, depending on the situation, either removing `StopTime` or marking the model `partial` could fulfill this need.

Reply:
I added the non-partial restriction, and inheritance of experiment is in #2535